src/Features/Core/Portable/SplitOrMergeIfStatements/AbstractMergeIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,86 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeIfStatementsCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        protected abstract bool IsApplicableSpan(SyntaxNode node, TextSpan span, out SyntaxNode ifLikeStatement);
+
+        protected abstract CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument, string ifKeywordText);
+
+        protected abstract Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifLikeStatement, CancellationToken cancellationToken);
+
+        protected abstract SyntaxNode GetChangedRoot(Document document, SyntaxNode root, SyntaxNode ifLikeStatement);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            if (IsApplicableSpan(node, context.Span, out var ifLikeStatement))
+            {
+                var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+                var syntaxKinds = context.Document.GetLanguageService<ISyntaxKindsService>();
+
+                if (await CanBeMergedAsync(context.Document, ifLikeStatement, context.CancellationToken).ConfigureAwait(false))
+                {
+                    context.RegisterRefactoring(
+                        CreateCodeAction(
+                            c => RefactorAsync(context.Document, context.Span, c),
+                            syntaxFacts.GetText(syntaxKinds.IfKeyword)));
+                }
+            }
+        }
+
+        private async Task<Document> RefactorAsync(Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(span, getInnermostNodeForTie: true);
+
+            Contract.ThrowIfFalse(IsApplicableSpan(node, span, out var ifLikeStatement));
+
+            var newRoot = GetChangedRoot(document, root, ifLikeStatement);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected static IReadOnlyList<SyntaxNode> WalkDownBlocks(ISyntaxFactsService syntaxFacts, IReadOnlyList<SyntaxNode> statements)
```

Comment:
WalkDownPureBlocks?

Reply:
➡️ Renamed in c9852b895b38ba50b4f51035566b7c13b21c62a7